nncf/data/dataset.py

```diff
@@ -73,6 +96,63 @@ def get_inference_data(self, indices: Optional[List[int]] = None) -> Iterable[Mo
         return DataProvider(self._data_source, self._transform_func, indices)
 
 
+class CountingDatasetWrapper(IDataset):
```

Comment:
I think it is better to extend `Evaluator` class instead of creation wrapper here. I suggest to use this class
```python
class IterationCounter:
    def __init__(self, iterable):
        self._iterable = iterable
        self._num_iterations = 0

    @property
    def num_iterations(self) -> int:
        return self._num_iterations

    def __iter__(self):
        for x in self._iterable:
            self._num_iterations += 1
            yield x
```
inside `Evaluator` class.

Also add property to `Evaluator` class that returns number of passed iterations during last validation process. For example,
```python
@property
def num_passed_iterations(self) -> int:
    """
    Returns number of passed iterations during last validation process.
    """
```

Reply:
In this case, I need to enter a flag that will enable or disable the collection of the number of iterations passed. Do you still support your proposal?

Reply:
Why not collect them all the time?

Reply:
It looks like it is not needed every time. I don't want to add overhead.

Reply:
I have updated the PR, please take a look.